src/DistributionsAD.jl

```diff
@@ -49,14 +43,39 @@ export TuringScalMvNormal,
 
 include("common.jl")
 include("univariate.jl")
+#=
 include("multivariate.jl")
 include("mvcategorical.jl")
 include("matrixvariate.jl")
 include("flatten.jl")
 include("arraydist.jl")
 include("filldist.jl")
-@init @require ReverseDiff = "37e2e3b7-166d-5795-8a7a-e32c996b4267" begin
-    include("reversediff.jl")
+=#
+function __init__()
+    @require ForwardDiff="f6369f11-7733-5829-9624-2563aa707210" begin
+        using ForwardDiff: @define_binary_dual_op # Needed for `eval`ing diffrules here
+        include("forwarddiff.jl")
+    end
+
+    @require ReverseDiff = "37e2e3b7-166d-5795-8a7a-e32c996b4267" begin
+        include("reversediff.jl")
+    end
+
+    @require Tracker="9f7883ad-71c0-57eb-9f7f-b5c9e6d3789c" begin
+        using DiffRules
+        using ZygoteRules
+        using SpecialFunctions
+        using LinearAlgebra: AbstractTriangular
```

Comment:
I guess you have to nest dependencies here (see https://github.com/JuliaPackaging/Requires.jl#multiple-requirements), since they are/should not be dependencies of DistributionsAD. Unfortunately, currently some adjoints for Tracker use the pullback of Zygote internally (I assume for performance reasons, I haven't worked on this implementation), indicated by the use of ZygoteRules here. However, that means that users will get cryptic error messages if they don't load Zygote (see e.g. https://github.com/TuringLang/Bijectors.jl/issues/119#issuecomment-649639106) - maybe we could try to remove the dependency on Zygote(Rules) all together and rewrite these adjoints in a different, hopefully not too inefficient way.

Reply:
Hmm, if we nest the dependencies, the user would have to load all of the by hand, which I though would be a bit cumbersome. If we make them dependencies of DistributionsAD (which is not very nice, I agree) that would not be the case.

We could get rid of the Zygote(Rules) dependency like this:
```julia
 @grad function turing_chol(A::AbstractMatrix, check)                                                                                                                                                                                                                                        
    C = cholesky(data(A), check=check)                                                                                                                                                                                                                                                      
    L = C.L                                                                                                                                                                                                                                                                                 
    ϕ(A::AbstractMatrix) = UpperTriangular(triu(A,1)) .+ Diagonal(A) ./ 2                                                                                                                                                                                                                   
    function back(L̄)                                                                                                                                                                                                                                                                        
        P = ϕ(L̄*L)                                                                                                                                                                                                                                                                          
        L1 = inv(L)                                                                                                                                                                                                                                                                         
        Ā = ϕ(L1' * (P .+ P') * L1)                                                                                                                                                                                                                                                         
        (Ā,nothing)                                                                                                                                                                                                                                                                         
    end                                                                                                                                                                                                                                                                                     
    return (C.factors, C.info), Δ -> back(data(Δ[1]))                                                                                                                                                                                                                                       
end  
```
Code based on [this](https://homepages.inf.ed.ac.uk/imurray2/pub/16choldiff/choldiff.pdf)

Reply:
It seems ChainRules (not ChainRulesCore) actually contains some implementations of this paper for the reverse-mode case: https://github.com/JuliaDiff/ChainRules.jl/blob/224e553132f26979e7f5c34fccaf769788022299/src/rulesets/LinearAlgebra/factorization.jl#L72-L83

This doesn't solve the issues here with mixed AD but it seems that all the special adjoints and workarounds for Cholesky with ZygoteRules/Zygote might not be needed at all for the Zygote backend. One idea might be to depend on ChainRules for now (although ideally we would only want to depend on ChainRulesCore), and then call `rrule` here explicitly, without having to rely on Zygote. I still think it's weird that users should have to load both AD backends for these adjoints.

Reply:
Yes, I also think they should only have to load `Tracker` here. I could copy the reverse functions from ChainRules and use them in the `@grad` adjoint for Tracker?

Reply:
I usually try to avoid code duplications, if not needed. IMO it would be fine to add a general dependency on ChainRules (its dependencies don't seem to be too bad, see https://github.com/JuliaDiff/ChainRules.jl/blob/224e553132f26979e7f5c34fccaf769788022299/Project.toml#L5-L10) instead of copying the code. Then these rules and adjoints are guaranteed to be available for every backend, and hopefully we could call `rrule` here directly without having to rely on Zygote (see https://www.juliadiff.org/ChainRulesCore.jl/stable/#Example-of-using-ChainRules-directly.-1 for some examples).

Reply:
I thought `Tracker` does not support `ChainRules` yet?

Reply:
Ahh, yes, sorry, we can just call the `rrule` directly. ofc :)

Reply:
Yes, but that's a different use case here. Tracker doesn't support it just means does Tracker can't process the custom adjoints defined with ChainRulesCore. However, you could still just call `rrule(cholesky, ...)` inside of the custom adjoint for Tracker and would get the function and its pullback, similar to what is currently done with ZygoteRules/Zygote.

Reply:
Ah I saw your last reply only after I had sent my comment :)

Reply:
My attempt for both `Tracker` and `ReverseDiff` is in `common.jl`, `tracker.jl`, and `reversediffx.jl`